src/main/java/httpServer/ListenerAndResponder.java

```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
-            printer.println(message);
-
+    private void printServerResponse(String httpResponse, Socket clientSocket) throws IOException {
+        ClientWriteable printer = clientWriterFactory.makePrinter(clientSocket);
+        System.out.println(httpResponse);
+        printer.print(httpResponse);
+        printer.close();
     }
 
-    private String readAllLines(ClientReadable bufferedReader)  {
-        String line;
-        StringBuilder stringBuilder = new StringBuilder();
+    private HTTPRequest readAllLinesAndReturnHttpRequest(ClientReadable bufferedReader)  {
+
+        String character;
+        RequestBuilder requestBuilder = new RequestBuilder();
+        StringBuilder headers = new StringBuilder();
+        StringBuilder body = new StringBuilder();
+
         try {
-            while ((line = bufferedReader.readLine()) != null)
-            {
-                stringBuilder.append(line);
-                stringBuilder.append(CRLF);
-                if (line.equals("")) { break; }
+            while ((character = bufferedReader.read()) != null) {
+
+                headers.append(character);
+                if (checkIfReachedEndOfHeaders(headers)) {
+                    break;
+                }
             }
-            return stringBuilder.toString();
+
         } catch (IOException e) {
             e.printStackTrace();
-            return stringBuilder.toString();
+            return requestBuilder.build();
+        }
+
+        if (checkIfHeadersIndicateARequestBody(headers)) {
+            int bodyLength = getBodyLength(headers);
+            try {
+                while ((character = bufferedReader.read()) != null) {
+                    if (java.util.Objects.equals(bodyLength, 0)) {
+                        break;
+                    }
+                    body.append(character);
+                    if (body.length() == bodyLength) {
+                        break;
+                    }
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+                return requestBuilder.build();
+            }
+        }
+        return turnStringBuilderIntoHttpRequest(headers, body, requestBuilder);
+    }
+
+    private HTTPRequest turnStringBuilderIntoHttpRequest(StringBuilder headers, StringBuilder body, RequestBuilder requestBuilder) {
+        String[] headersArray = headers.toString().split(crlf);
+        for (String header: headersArray) {
+            requestBuilder = discernHeaderTypeAndAddToRequestBuilder(header, requestBuilder);
+        }
+        if (body.length() != 0) {
+            requestBuilder.setBody(body.toString());
         }
+        return requestBuilder.build();
     }
+
+    private RequestBuilder discernHeaderTypeAndAddToRequestBuilder(String line, RequestBuilder requestBuilder) {
+        if (!line.contains(": ")) {
+            requestBuilder.buildRequestLine(line);
+        } else {
+            requestBuilder.buildHeaderLine(line);
+
+        }
```

Comment:
I think it will be more straightforward to follow if you do:
```java
if(line.contains(": ") {
    requestBuilder.buildHeaderLine(line)
} else {
    requestBuilder.buildRequestLine(line)\
}
```